vllm/v1/engine/core.py

```diff
@@ -242,6 +242,10 @@ def step_with_batch_queue(self) -> Optional[EngineCoreOutputs]:
 
         # If no more requests can be scheduled and the job queue is not empty,
         # block until the first batch in the job queue is finished.
+        # TODO(comaniac): Ideally we should peek the first batch in the
+        # job queue to check if it's finished before scheduling a new batch,
+        # but peeking the first element in a queue is not thread-safe,
+        # so we need more work.
```

Comment:
Actually, I don't fully understand the comment here. Could you please elaborate more?

Reply:
The ideal case is after each scheduling step, we peek the first batch and check if it's done. If it's done then we process its output immediately; otherwise we skip it and schedule the next batch. However, we use deque to maintain flying batches, so its thread safe and doesn't support peeking, so we cannot achieve the above behavior.